File: protokimica-portal/backend/app/core/modulos.py
"""
Qué módulos puede abrir cada quien.

Hasta ahora el portal solo controlaba la ESCRITURA (`solo_lectura_no`) y la
visibilidad por área dentro de Master Planner. Cualquier usuario autenticado
podía leer cualquier módulo — un agente de Logística veía todos los
indicadores de la empresa.

Esto se aplica en el backend, no escondiendo el menú: esconder un botón no
impide escribir la URL a mano.

Regla general: **el rol decide a qué módulo entras, el área decide qué ves
dentro.** Un líder entra a Indicadores, pero solo ve los de su área.
"""
from fastapi import Depends, HTTPException, status

from app.core.deps import get_current_user
from app.models.user import User

# modulo -> roles que pueden abrirlo. `admin` va en todos por definición.
ACCESO_POR_MODULO: dict[str, set[str]] = {
    "inicio": {"admin", "gerencia", "lider", "agente", "lectura"},
    "pqrs": {"admin", "gerencia", "lider", "agente", "lectura"},
    "master_planner": {"admin", "gerencia", "lider", "agente", "lectura"},
    # Un agente organiza su trabajo en PQRS y Master Planner; los indicadores
    # son de quien responde por ellos. El líder entra porque le toca registrar
    # los de su área cada mes.
    "indicadores": {"admin", "gerencia", "lider"},
    # Las encuestas quedan abiertas como PQRS: la satisfacción del cliente la
    # consulta cualquiera que atienda. Si un día hay que cerrarlas, se cambia
    # aquí y no en veinte endpoints.
    "encuestas": {"admin", "gerencia", "lider", "agente", "lectura"},
    "admin": {"admin"},
}

ETIQUETAS = {
    "inicio": "Inicio",
    "pqrs": "PQRS",
    "master_planner": "Master Planner",
    "indicadores": "Indicadores",
    "encuestas": "Encuestas",
    "admin": "Administración",
}


def puede_ver_modulo(usuario: User, modulo: str) -> bool:
    permitidos = ACCESO_POR_MODULO.get(modulo)
    if permitidos is None:
        raise ValueError(f"Módulo desconocido: '{modulo}'")
    return usuario.rol in permitidos


def modulos_de(usuario: User) -> list[str]:
    """Los módulos que este usuario puede abrir, en orden de menú."""
    return [m for m in ACCESO_POR_MODULO if puede_ver_modulo(usuario, m)]


def requiere_modulo(modulo: str):
    """
    Dependencia de FastAPI: corta el acceso a un módulo completo.

    Uso:  _: User = Depends(requiere_modulo("indicadores"))
    """
    def verificar(current_user: User = Depends(get_current_user)) -> User:
        if not puede_ver_modulo(current_user, modulo):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=(
                    f"Tu rol no tiene acceso al módulo de {ETIQUETAS.get(modulo, modulo)}. "
                    "Si necesitas entrar, pídeselo a un administrador."
                ),
            )
        return current_user
    return verificar


# ── Alcance dentro de Indicadores ─────────────────────────────

def ve_todos_los_indicadores(usuario: User) -> bool:
    """
    Gerencia y admin ven la empresa completa. Un líder ve los indicadores de
    su área: son los que le toca responder y registrar.
    """
    return usuario.rol in ("admin", "gerencia")

File: protokimica-portal/frontend/src/core/modulos.js
/**
 * Qué módulos puede abrir cada rol — gemelo de `backend/app/core/modulos.py`.
 *
 * Vive dos veces porque el frontend no puede importar Python. Una prueba
 * (`tests/modulos.test.mjs`) verifica que las dos listas digan lo mismo.
 *
 * Esto es SOLO para no mostrar un menú que no lleva a ninguna parte. El
 * bloqueo de verdad está en el backend: esconder un botón no impide escribir
 * la URL a mano.
 *
 * Regla general: **el rol decide a qué módulo entras, el área decide qué ves
 * dentro.** Un líder entra a Indicadores, pero solo ve los de su área.
 */
export const ACCESO_POR_MODULO = {
  inicio: ['admin', 'gerencia', 'lider', 'agente', 'lectura'],
  pqrs: ['admin', 'gerencia', 'lider', 'agente', 'lectura'],
  master_planner: ['admin', 'gerencia', 'lider', 'agente', 'lectura'],
  indicadores: ['admin', 'gerencia', 'lider'],
  // La mejora es trabajo de los líderes de área. Gerencia queda fuera: el
  // avance se le reporta, no es un tablero más que mirar.
  mejora: ['admin', 'lider'],
  encuestas: ['admin', 'gerencia', 'lider', 'agente', 'lectura'],
  admin: ['admin'],
}

/** Ruta de cada módulo, para filtrar el menú y proteger las rutas. */
export const RUTA_DE_MODULO = {
  inicio: '/',
  pqrs: '/pqrs',
  master_planner: '/master-planner',
  indicadores: '/indicadores',
  mejora: '/mejora',
  encuestas: '/encuestas',
  admin: '/admin',
}

export function puedeVerModulo(usuario, modulo) {
  const permitidos = ACCESO_POR_MODULO[modulo]
  if (!permitidos) return false
  return permitidos.includes(usuario?.rol)
}

/** Los módulos que este usuario puede abrir, en orden de menú. */
export function modulosDe(usuario) {
  return Object.keys(ACCESO_POR_MODULO).filter(m => puedeVerModulo(usuario, m))
}

/**
 * A qué módulo pertenece una ruta. `/pqrs/12` es del módulo `pqrs`.
 * Devuelve null si la ruta no está mapeada — esas no se bloquean, porque
 * bloquear por omisión dejaría fuera pantallas nuevas sin avisar.
 */
export function moduloDeRuta(ruta) {
  const limpia = (ruta || '/').split('?')[0]
  if (limpia === '/' || limpia === '') return 'inicio'
  const encontrado = Object.entries(RUTA_DE_MODULO)
    .filter(([, r]) => r !== '/')
    .find(([, r]) => limpia === r || limpia.startsWith(`${r}/`))
  return encontrado ? encontrado[0] : null
}

/**
 * A dónde mandar a alguien que llegó a un módulo que no le corresponde.
 * Al inicio, que todos pueden ver — nunca a una pantalla que también le
 * cerraría la puerta, porque eso sería un ciclo de redirecciones.
 */
export const RUTA_POR_DEFECTO = '/'
